### PR TITLE
Adding peek support for endpoints

### PR DESCRIPTION
**Pull request by jherico**
This allows the creation of mappings that don't inhibit other mappings using the same source endpoints.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: e2d84bf472185a7604948957845e3421bcc1d991 available at: https://gist.github.com/5ef2775a2d6ad1cb2800<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6250.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6250.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 80bbd64f2f94ee4f2547bb8f67d2cc47a85dfba1 available at: https://gist.github.com/5ef2775a2d6ad1cb2800<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6250.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6250.dmg

File: libraries/controllers/src/controllers/impl/endpoints/AnyEndpoint.cpp [hunk 1]
```diff
@@ -27,6 +27,17 @@ AnyEndpoint::AnyEndpoint(Endpoint::List children) : Endpoint(Input::INVALID_INPU
     }
 }
 
+float AnyEndpoint::peek() const {
+    for (auto& child : _children) {
+        float childResult = child->peek();
+        if (childResult != 0.0f) {
+            return childResult;
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/impl/endpoints/AnyEndpoint.cpp]: this logic is different than the value logic... do you really want it to be?

**Comment by ZappoMan** [comment]
needs update from master before it can be merged

File: libraries/controllers/src/controllers/impl/endpoints/AnyEndpoint.cpp (hunk 1 above)
Comment (jherico) [inline libraries/controllers/src/controllers/impl/endpoints/AnyEndpoint.cpp]: yes.  For AnyEndpoint, we want to query the value of *every* child so the side effects of calling value (i.e. marking the endpoint as already read this frame) take effect.  The point of peek is to not have any side effects, so we can return the first non-zero value we see without calling the rest.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 446d424e12581a33d4d4f45dab33e9751e93e6fc available at: https://gist.github.com/5ef2775a2d6ad1cb2800<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6250.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6250.dmg

File: libraries/controllers/src/controllers/impl/endpoints/AnyEndpoint.cpp [hunk 2]
```diff
@@ -27,13 +27,19 @@ AnyEndpoint::AnyEndpoint(Endpoint::List children) : Endpoint(Input::INVALID_INPU
     }
 }
 
+float AnyEndpoint::peek() const {
+    float result = 0;
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/impl/endpoints/AnyEndpoint.cpp]: should be 0.0f or will produce warning

File: libraries/controllers/src/controllers/impl/endpoints/AnyEndpoint.cpp [hunk 3]
```diff
@@ -27,13 +27,19 @@ AnyEndpoint::AnyEndpoint(Endpoint::List children) : Endpoint(Input::INVALID_INPU
...
+    for (auto& child : _children) {
+        result = std::max(result, child->peek());
+    }
+    return result;
+}
+
+// Fetching the value must trigger any necessary side effects of value() on ALL the children.
 float AnyEndpoint::value() {
     float result = 0;
```
Comment (ZappoMan) [inline libraries/controllers/src/controllers/impl/endpoints/AnyEndpoint.cpp]: should be 0.0f or will produce warning

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9e3a9447cdd391f04b6b3fe51f9af929f7c6ca9c available at: https://gist.github.com/5ef2775a2d6ad1cb2800<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6250.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6250.dmg

**Comment by ZappoMan** [comment]
handControllerGrab.js does not appear to work with this PR - I'm not seeing any syntax errors, it's just not working.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 79135d37309449a8bb148098844f6cdc44e432ac available at: https://gist.github.com/5ef2775a2d6ad1cb2800<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6250.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6250.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0e13348b5a3f4874552097cff8f58a122be72676 available at: https://gist.github.com/5ef2775a2d6ad1cb2800<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6250.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6250.dmg

**Pull request merged**